library/std/src/sys/unix/futex.rs

```diff
@@ -4,33 +4,45 @@
     all(target_os = "emscripten", target_feature = "atomics")
 ))]
 
-#[cfg(any(target_os = "linux", target_os = "android"))]
-use crate::convert::TryInto;
-#[cfg(any(target_os = "linux", target_os = "android"))]
-use crate::ptr::null;
 use crate::sync::atomic::AtomicI32;
 use crate::time::Duration;
 
 #[cfg(any(target_os = "linux", target_os = "android"))]
 pub fn futex_wait(futex: &AtomicI32, expected: i32, timeout: Option<Duration>) -> bool {
-    let timespec = timeout.and_then(|d| {
-        Some(libc::timespec {
-            // Sleep forever if the timeout is longer than fits in a timespec.
-            tv_sec: d.as_secs().try_into().ok()?,
-            // This conversion never truncates, as subsec_nanos is always <1e9.
-            tv_nsec: d.subsec_nanos() as _,
-        })
-    });
-    let r = unsafe {
-        libc::syscall(
-            libc::SYS_futex,
-            futex as *const AtomicI32,
-            libc::FUTEX_WAIT | libc::FUTEX_PRIVATE_FLAG,
-            expected,
-            timespec.as_ref().map_or(null(), |d| d as *const libc::timespec),
-        )
-    };
-    !(r < 0 && super::os::errno() == libc::ETIMEDOUT)
+    use super::time::Instant;
+    use crate::ptr::null;
+    use crate::sync::atomic::Ordering::Relaxed;
+
+    // Calculate the timeout as an absolute timespec.
+    let timespec =
+        timeout.and_then(|d| Some(Instant::now().checked_add_duration(&d)?.as_timespec()));
```

Comment:
Futex used CLOCK_MONOTONIC by default. https://github.com/rust-lang/rust/pull/88714 would switch Instant to CLOCK_BOOTTIME.

Reply:
Good catch!

I've made the internal Timespec available to the rest of sys::unix in https://github.com/rust-lang/rust/pull/95035/commits/23badeb4cb22afd09e0264348800e44cda80af38, and then used that to explicitly get a CLOCK_MONOTONIC timespec in futex_wait in https://github.com/rust-lang/rust/pull/95035/commits/c9ae3fe68fc4477350476acb649b817a80bfd973.